src/shared.rs

```diff
@@ -0,0 +1,211 @@
+    fn new(error: E) -> Self {
+        SharedError { error: Arc::new(error) }
+    }
+}
+
+struct SyncedInner<F>
+    where F: Future
+{
+    original_future: F,
+    result: Option<Result<Async<SharedItem<F::Item>>, SharedError<F::Error>>>,
+    tasks_receiver: Mutex<Receiver<Task>>,
+}
+
+struct Inner<F>
+    where F: Future
+{
+    synced_inner: RwLock<SyncedInner<F>>,
+    tasks_unpark_started: AtomicBool,
+}
+
+/// TODO: doc
+#[must_use = "futures do nothing unless polled"]
+pub struct Shared<F>
+    where F: Future
+{
+    inner: Arc<Inner<F>>,
+    tasks_sender: Sender<Task>,
+}
+
+pub fn new<F>(future: F) -> Shared<F>
+    where F: Future
+{
+    let (tasks_sender, tasks_receiver) = channel();
+    Shared {
+        inner: Arc::new(Inner {
+            synced_inner: RwLock::new(SyncedInner {
+                original_future: future,
+                result: None,
+                tasks_receiver: Mutex::new(tasks_receiver),
+            }),
+            tasks_unpark_started: AtomicBool::new(false),
+        }),
+        tasks_sender: tasks_sender,
+    }
+}
+
+impl<F> Future for Shared<F>
+    where F: Future
+{
+    type Item = SharedItem<F::Item>;
+    type Error = SharedError<F::Error>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let mut polled_result: Option<Result<Async<SharedItem<F::Item>>, SharedError<F::Error>>> =
+            None;
+
+        // If the result is ready, just return it
+        match self.inner.synced_inner.try_read() {
+            Ok(inner_guard) => {
+                let ref inner = *inner_guard;
+                if let Some(ref result) = inner.result {
+                    return result.clone();
+                }
+            }
+            _ => {} // Mutex is locked for write
+        }
+
+        // The result was not ready.
+        match self.inner.synced_inner.try_write() {
+            Ok(mut inner_guard) => {
+                let ref mut inner = *inner_guard;
+
+                // By the time that synced_inner was unlocked, other thread could poll the result,
+                // so we check if result has a value
+                if inner.result.is_some() {
+                    polled_result = inner.result.clone();
+                } else {
+                    match inner.original_future.poll() {
+                        Ok(Async::Ready(item)) => {
+                            inner.result = Some(Ok(Async::Ready(SharedItem::new(item))));
+                            polled_result = inner.result.clone();
+                        }
+                        Err(error) => {
+                            inner.result = Some(Err(SharedError::new(error)));
+                            polled_result = inner.result.clone();
+                        }
+                        Ok(Async::NotReady) => {} // Will be handled later
+                    }
+                }
+            }
+            Err(_) => {} // Will be handled later
+        }
+
+        if polled_result.is_some() {
```

Comment:
This could be `if let Some(...) = polled_result`, removing the `unreachable!` below I believe.